### ISSUE TITLE
official release

### ISSUE THREAD

**Issue by brucou** [Heydon/bruck#15]
Hi Heydon,

I could not find any release on this github directory. Do you plan an official release anytime soon?

Thanks and big up for the great library.

**Comment by Heydon** [comment]
@brucou Hi! Thank you for the interest.

My plans before making an official release:

* Give each component a render function so that it can rerender on prop changes
* Include `<s-witcher>` (based on the [Flexbox Holy Albatross](http://www.heydonworks.com/article/the-flexbox-holy-albatross))
* Document each component properly with jsdoc
* Make sure each component mirrors attributes and props, so it can be altered imperatively

I'd be interested in your thoughts on what else might be good to have.

**Comment by brucou** [comment]
Hi Heydon,

I [covered the current Bruck features](https://www.infoq.com/news/2019/02/bruck-quick-layout-prototyping?utm_source=infoq&utm_campaign=user_page&utm_medium=link). I was asking about the final release as a part of my article. I'll be happy to cover that release when it occurs. That's my comment as a tech writer.

As a practitioner, I would say to remain true to the lean and mean 'Quickly create and comment on interface layouts.' approach. So if there is a common use case when render functions etc. makes it faster to create layouts vs. not having them, I guess it works fine. I would also like to see more examples of actual layouts prototyped with Bruck used in some nice known web page (somehow featuring fake data, and actual data). The features are already great as is, I think it is a matter as you said of packaging it nicely to the target audience. That means doc, of course and ideally examples, which fit into a designer's process (lo-fi first prototype, feedback+refinement, hi-fi, etc).

Of course it is easier to give advice, than to actually do anything, so feel free to disregard any of these suggestions if they do not fit your vision.

**Comment by Heydon** [comment]
@brucou no, that's helpful feedback. Will bear your comments in mind. The render function stuff shouldn't affect performqnce much, but it will mean users could include Bruck components inside Preact or Vue (etc) and passing props would affect the internal component. I want it to be a standalone thing, but also play nicely with SPA frameworks, so it's possible to also make working 'dummy' apps. Hope that makes sense.

**Comment by brucou** [comment]
It does make sense. That means though your target audience includes developers (which I am a part of). I was expecting it to be more geared towards designers, and I would not expect designers to know javascript, vue or else, those being a distinct skill set. I mean giving the time you can waste with build and miscellanous issues in the front-end, I would expect designers to keep safe to tools they know. Now if your audience are developers which are not so good in design (better at using vue, react, js than sketch), then that does make sense. Now that I think about it, designers might already know how to produce prototypes quickly with the usual prototyping tools (balsamiq, etc.). For me for instance who is a dummy at design, reactive web components may make me win time for sure. The alternative would be to use a UI library and learn its layout component, its component set and class parameterization and the usual many quirks and then find a way to generate random data. Not like it is impossible, but it is time consuming. Anyways, having to write no code to do anything is always the better option when that is available. 

In short, I guess it is all about which users and use cases you want to facilitate.

Last thing, it is probably not an issue, as for prototyping you can use any modern browsers of your choice, so houdini and web component, custom elements etc. is fine. I just wanted to let you now about this library which allow you to do web components without having the specifications implemented, i.e. that would work down to IE9. It is called [regular elements](https://github.com/WebReflection/regular-elements) and should work down to IE9 and is 2kb long. [wicked elements](https://github.com/WebReflection/wicked-elements) is the same as regular elements but with a render function.

**Comment by brucou** [comment]
Hi Heydon, how have you been doing? Coming back to you to see what is up with the project. I also wanted to cover `https://every-layout.dev/` just as I did for bruck. However I see it is for pay with exceptions. I would qualify for at least one of those exceptions (the minority part), but my interest is primarily to write an article on it, so I thought I should ask you directly here.

**Comment by Heydon** [comment]
@brocou Hi, sorry, have not been looking at this directly for a while. Regular elements look really interesting. One thing I don't understand: how do I add props (and typings for them)? In a custom element class definition I'd use getters and setters, like:

```js
  get height() {
    return this.getAttribute('height') || 'auto';
  }

  set height(val) {
    return this.setAttribute('height', val);
  }
```

**Comment by brucou** [comment]
Hi @Heydon,

you don't actually define a getter or setter. You define an `onattributechanged` callback which is called anytime you change an attribute of the element. Changing an attribute of an element can be done as usual with `el.setAttribute`. 

The great thing about this is that the code is about 100LOC, so you can have a look to remove any doubt about its behaviour: https://github.com/WebReflection/regular-elements/blob/master/esm/index.js

It basically uses `MutationObserver` to identify when a given selector appears on the DOM. This could be `button.alive` or `#complex > selector [data-available]`. You associate a *regular element* to that selector, so when the selector appears, it runs the `onconnected` callback. You can specify a list of attributes whose changes you want to react to. Your element is a regular DOM element, so you can change its attributes with whatever standard way. For instance `.setAttribute`. When you do that, the `onattributechanged` callback will be called to do any kind of reactive processing you want to do. All the same if you want to read the value of an attribute, you do that the usual way (for instance `.getAttribute`). So the point is you do not have a class which proxies getter and setter for the DOM element, you directly access the DOM element to get and set attributes. That means you can perfectly set an attribute that you are not meant to. But it won't matter because if you configure `attributeFilter` correctly, it will ignore that unexpected attribute.

So this is pretty low-level, but then you can build anything you want on top of it (default values, etc.). [wicked-elements](https://github.com/WebReflection/wicked-elements) for instance adds a `render` function to that so that the DOM element `innerhtml` is updated when an attribute changes. What I don't know about is typing though. I guess you can just write a `.ts` version of the library.

Also wanted to add that this will be compatible with any framework you may want to use because they will always end up modifying an attribute with the DOM API call `.setAttribute`. If the library was higher level, you might have to write an adapter to integrate with specific libraries. Here you don't. So you have a web component functionality that is supported by I think it was up to IE9.
